### PR TITLE
[jit] Module clone work with shared ClassType

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32206 [quant] Re-enable test_nested that has different qconfig for shared ClassType
* #32205 [quant][graphmode] Support quantizing shared ClassType with different qconfigs
* #32204 [quant][graphmode] Default to non-inplace in graph mode quantization API
* #32203 Adding QConfigTypePtrMap
* #32202 [refactor] Move some of the helper functions for public use
* #31408 [quant][graphmode] `insert_quant_dequant` pass support shared class types
* **#31970 [jit] Module clone work with shared ClassType**

Summary:
Now that the ClassType can be shared among different module instances, we'll
preserve the sharing in clone as well, that is if the original module has
a ClassType that is shared, we'll clone this ClassType once and share it between
different module instances as well.

Test Plan:
build/test/test_jit

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D19406251](https://our.internmc.facebook.com/intern/diff/D19406251)

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 86fa90e9:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=dedd16b4181cae81e37e978cd3bf24c1ba35ca05&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base dedd16b4 since Jan 08
  <details>
  <summary>
  You may want to rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=86fa90e9f50e7c61fad09a061083eeb7556771a7).

### 1 failure *not* recognized by patterns:
| Job | Step | Status |
| --- | --- | --- |
| CircleCI <sup>pytorch_xla_linux_xenial_py3_6_clang7_build</sup> | <sup>Build</sup> | &#128721; [Broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=dedd16b4181cae81e37e978cd3bf24c1ba35ca05&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35&highlight_job=pytorch_xla_linux_xenial_py3_6_clang7_build) |


---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Comment by ZolotukhinM** [comment]
We probably need two versions of `clone`: one that creates a new type and one that uses the old one. E.g. when we clone a module in order to quantize it, we need a new type as we're going to mutate its methods. Inside quantization, however, where we have several stages (insert observers, replace observers with quant-dequant, etc.), when we need a copy of the module we would clone without cloning the type (since we've already created the new type in the very beginning).

**Comment by jerryzh168** [comment]
What is clone without cloning type? are you talking about clone_instance or just don't clone?

**Comment by jerryzh168** [comment]
I think insert_quant_dequant should work without cloning at all, the reason we clone is to invalidate the executor cache, since after insert_observer we'll run the model and the executor is cached.

**Comment by ZolotukhinM** [comment]
> I think insert_quant_dequant should work without cloning at all, the reason we clone is to invalidate the executor cache, since after insert_observer we'll run the model and the executor is cached.

We would still need to do that, right? If we don't create a new type at that moment, an old graph (with observers rather than with q-dq nodes) will be executed otherwise.

**Comment by jerryzh168** [comment]
> > I think insert_quant_dequant should work without cloning at all, the reason we clone is to invalidate the executor cache, since after insert_observer we'll run the model and the executor is cached.
> 
> We would still need to do that, right? If we don't create a new type at that moment, an old graph (with observers rather than with q-dq nodes) will be executed otherwise.

Yeah, that's what we do right now, we'll clone in both insert_observers and insert_quant_dequant

**Comment by jerryzh168** [comment]
We also discovered that the original clone does not work anymore after the shared ClassType change.
If a module contains two instances that shares the same type, e.g.
```
class M(torch.nn.Module):
    def __init__(self):
        super(M, self).__init__()
        self.fc1 = torch.nn.Linear(5, 5)
        self.fc2 = torch.nn.Linear(5, 5)
    
    def forward(self, x):
           return self.fc1(self.fc2(x))
a = torch.jit.script(M())
a = a.copy()
print(a._c.dump_to_str())
```
they will have different type in the module but same type in the forward function:
```
module __torch__.___torch_mangle_268.M {
  parameters {
  }
  attributes {
    training = True
    fc1 = <__torch__.torch.nn.modules.linear.___torch_mangle_269.Linear object at 0x7fe2a0658870>
    fc2 = <__torch__.torch.nn.modules.linear.___torch_mangle_270.Linear object at 0x7fe2a0658640>
  }
  methods {
    method forward {
      graph(%self : __torch__.___torch_mangle_268.M,
            %x.1 : Tensor):
        %2 : __torch__.torch.nn.modules.linear.___torch_mangle_270.Linear = prim::GetAttr[name="fc1"](%self)
        %3 : __torch__.torch.nn.modules.linear.___torch_mangle_270.Linear = prim::GetAttr[name="fc2"](%self)
        %4 : Tensor = prim::CallMethod[name="forward"](%3, %x.1) # <ipython-input-67-70905ca9c75f>:8:27
        %5 : Tensor = prim::CallMethod[name="forward"](%2, %4) # <ipython-input-67-70905ca9c75f>:8:18
        return (%5)
  
    }
  }
```

File: test/test_quantization.py [hunk 1]
```diff
@@ -905,6 +905,11 @@ def test_conv_bn(self):
         result_script = model_script(self.img_data[0][0])
         self.assertEqual(result_eager, result_script)
 
+    @unittest.skip(
+        "Temporarily skip the test since we don't have"
+        "support for different quantization configurations for shared"
+        "ClassType right now, sub1.fc and fc3 shares the ClassType but for"
+        " sub1.fc qconfig is None, and fc3 is quantized with default_qconfig")
```
Comment (ZolotukhinM) [inline test/test_quantization.py]: Was this test working correctly before? If not, it's better to split this change into a separate commit.

**Review by ZolotukhinM — APPROVED** [review]
Looks good, please find a couple of comments inline!

File: test/test_quantization.py (hunk 1 above)
Comment (jerryzh168) [reply test/test_quantization.py]: yeah, this is working before since we clone will create new type for all types so we don't have type sharing after clone previously.
